presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1808,53 +1809,151 @@ private Scope computeAndAssignOrderByScopeWithAggregation(OrderBy node, Scope so
                     }
+                    throw new SemanticException(COLUMN_NAME_NOT_SPECIFIED, allColumns, "SELECT * not allowed from relation that has no columns");
+                }
+
+                analyzeAllColumnsFromTable(fields, allColumns, node, scope, outputExpressionBuilder, relationType);
+            }
+        }
+
+        private void analyzeAllColumnsFromTable(List<Field> fields, AllColumns allColumns, QuerySpecification node, Scope scope, ImmutableList.Builder<Expression> outputExpressionBuilder, RelationType relationType)
+        {
+            ImmutableList.Builder<Field> itemOutputFieldBuilder = ImmutableList.builder();
+
+            for (Field field : fields) {
+                int fieldIndex = relationType.indexOf(field);
+                FieldReference fieldReference = new FieldReference(fieldIndex);
+                outputExpressionBuilder.add(fieldReference);
+                itemOutputFieldBuilder.add(Field.newUnqualified(field.getName(), field.getType(), field.getOriginTable(), field.getOriginColumnName(), false));
+                ExpressionAnalysis expressionAnalysis = analyzeExpression(fieldReference, scope);
+
+                Type type = expressionAnalysis.getType(fieldReference);
+                if (node.getSelect().isDistinct() && !type.isComparable()) {
+                    throw new SemanticException(TYPE_MISMATCH, node.getSelect(), "DISTINCT can only be applied to comparable types (actual: %s)", type);
+                }
+            }
+            analysis.setSelectAllResultFields(allColumns, itemOutputFieldBuilder.build());
+        }
+
+        private void analyzeAllFieldsFromRowTypeExpression(Expression expression, AllColumns allColumns, QuerySpecification node, Scope scope, ImmutableList.Builder<Expression> outputExpressionBuilder)
+        {
+            ImmutableList.Builder<Field> itemOutputFieldBuilder = ImmutableList.builder();
+
+            ExpressionAnalysis expressionAnalysis = analyzeExpression(expression, scope);
+            Type type = expressionAnalysis.getType(expression);
+            if (!(type instanceof RowType)) {
+                throw new SemanticException(INVALID_ALL_COLUMNS_TARGET, node.getSelect(), "SELECT * can only be applied to expression of Row type");
+            }
+            if (!allColumns.getAliases().isEmpty()) {
+                validateColumnAliases(allColumns.getAliases(), ((RowType) type).getFields().size());
+            }
+            analysis.recordSubqueries(node, expressionAnalysis);
+
+            for (int i = 0; i < ((RowType) type).getFields().size(); i++) {
+                Expression outputExpression = new SubscriptExpression(expression, new LongLiteral("" + (i + 1)));
+                ExpressionAnalysis outputExpressionAnalysis = analyzeExpression(outputExpression, scope);
+                outputExpressionBuilder.add(outputExpression);
```

Comment:
This can be a problem for non-deterministic or expensive expressions, since it's inlining the expression into each of the expanded select items. This should be handled during planning by adding projection that dereferences each field.

Reply:
Yes, the expression is inlined in each expanded select item, but only 'syntactically'. The Planner identifies all occurrences as the same expression (via `TranslationMap`), and it is executed once.

Reply:
I ran `EXPLAIN` on
`SELECT (SELECT (min(regionkey), max(regionkey), true) FROM nation).*;`.
It only tablescaned once, computed the ROW from the subquery and then there was a Project which dereferenced each field. I think it is the plan you wanted.